mesonbuild/backend/backends.py

```diff
@@ -910,7 +910,12 @@ def _determine_ext_objs(self, extobj: 'build.ExtractedObjects', proj_dir_to_buil
 
         for osrc in sources:
             objname = self.object_filename_from_source(extobj.target, osrc)
-            objpath = os.path.join(proj_dir_to_build_root, targetdir, objname)
+            # Only add target private directories when not using Xcode backend.
+            # Xcode already has its own and will use that for objects.
+            if self.name != 'xcode':
+                objpath = os.path.join(proj_dir_to_build_root, targetdir, objname)
+            else:
+                objpath = os.path.join(proj_dir_to_build_root, objname)
             result.append(objpath)
```

Comment:
I think we can make this backend-agnostic by moving it to object_filename_from_source, which currently a 33-line function in backends.py, but also an independent 12-line function in xcodebackend.py.

We can add a kwarg to `def object_filename_from_source(self, target, source)` as `def object_filename_from_source(self, target, source, targetdir)`, and the xcode implementation can simply drop the target argument.

This would avoid creating the first instance where we check `self.name` against a backend name.

Reply:
I think such an idea would be best for a separate PR, unless you think it should be done in this one.

Reply:
@tangalbert919 please do it in this one, yes.

Reply:
I have implemented your idea, but I am uncertain if it is how you would've done it. I would like to emphasize that I would have preferred it to be in a separate PR, given that this affects all other backends, which is what I wanted to avoid.